pvlib/way_faster.py

```diff
@@ -0,0 +1,276 @@
+        and number of series cells ``Ns``
+    :returns: rough estimate of open circuit voltage [V]
+    """
+    # http://www.pveducation.org/pvcdrom/open-circuit-voltage
+    return nnsvt * np.log(il / io + 1.0)
+
+
+def bishop88(vd, il, io, rs, rsh, nnsvt):
+    """
+    Explicit calculation single-diode-model (SDM) currents and voltages using
+    diode junction voltages [1].
+
+    [1] "Computer simulation of the effects of electrical mismatches in
+        photovoltaic cell interconnection circuits" JW Bishop, Solar Cell (1988)
+        https://doi.org/10.1016/0379-6787(88)90059-2
+
+    :param vd: diode voltages [V}]
+    :param il: photo-generated current [A]
+    :param io: diode one reverse saturation or "dark" current [A]
+    :param rs: series resitance [ohms]
+    :param rsh: shunt resitance [ohms]
+    :param nnsvt" product of thermal voltage ``Vt`` [V], ideality factor ``n``,
+        and number of series cells ``Ns``
+    :returns: tuple containing currents [A], voltages [V], gradient ``di/dvd``,
+        gradient ``dv/dvd``, power [W], gradient ``dp/dv``, and gradient
+        ``d2p/dv/dvd``
+    """
+    a = np.exp(vd / nnsvt)
+    b = 1.0 / rsh
+    i = il - io * (a - 1.0) - vd * b
+    v = vd - i * rs
+    c = io * a / nnsvt
+    grad_i = - c - b  # di/dvd
+    grad_v = 1.0 - grad_i * rs  # dv/dvd
+    # dp/dv = d(iv)/dv = v * di/dv + i
+    grad = grad_i / grad_v  # di/dv
+    grad_p = v * grad + i  # dp/dv
+    grad2i = -c / nnsvt
+    grad2v = -grad2i * rs
+    grad2p = (
+        grad_v * grad + v * (grad2i/grad_v - grad_i*grad2v/grad_v**2) + grad_i
+    )
+    return i, v, grad_i, grad_v, i*v, grad_p, grad2p
+
+
+def newton_solver(fjx, x0, x, tol=EPS, damp=DAMP, log=False, test=True):
+    resnorm = np.inf  # nor of residuals
+    while resnorm > tol:
+        f, j = fjx(x0, *x)
+        newton_step = f / j
+        # don't let step get crazy
+        if np.abs(newton_step / x0) > damp:
+            break
+        x0 -= newton_step
+        resnorm = f**2
+        if log:
+            LOGGER.debug(
+                'x0=%g, newton step=%g, f=%g, resnorm=%g',
+                x0, newton_step, f, resnorm
+            )
+        if test:
+            f2, _ = fjx(x0 * (1.0 + DELTA), *x)
+            LOGGER.debug('test_grad=%g', (f2 - f) / x0 / DELTA)
+            LOGGER.debug('grad=%g', j)
+    return x0, f, j
+
+
+def slow_i_from_v(v, photocurrent, saturation_current, resistance_series,
+                  resistance_shunt, nNsVth):
+    """
+    This is a slow but reliable way to find current given any voltage.
+    """
+    # FIXME: everything is named the wrong thing!
+    il = photocurrent
+    io = saturation_current
+    rs = resistance_series
+    rsh = resistance_shunt
+    nnsvt = nNsVth
+    x = (il, io, rs, rsh, nnsvt)  # collect args
+    # first bound the search using voc
+    voc_est = est_voc(il, io, nnsvt)
+    vd = fminbound(lambda vd: (v - bishop88(vd, *x)[1])**2, 0.0, voc_est)
```

Comment:
It seems like the most straightforward option for current computation at specific terminal voltages might be to directly implement a vectorized, derivative-free brent algorithm with a sensible root bracket that will "guarantee" convergence. I am investigating better ways to vectorize, say, scipy's [brentq](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.brentq.html) for this purpose. It would be a shame to have to rewrite it from scratch just to get performant vectorization.

Reply:
I haven't reviewed all of this, but just want to suggest that simplicity should be a higher objective than speed for the reference implementation.

Reply:
Respectfully, I believe that what I'm suggesting is the simplest method that is _guaranteed_ to give you the most accurate answer given these governing equations:

1. Use an explicit method
2. Estimate open circuit voltage by assuming that shunt resistance is infinite
3. Bound the solution between `vd = 0` and `vd = voc_est` 
4. Use bisection, a "slow but stable" method, that is guaranteed to find the desired operating condition between the bounds

Reply:
Sorry, @mikofski , I was not commenting on your code at all but rather referring to @thunderfish24 suggestion to vectorize things.

Reply:
Ok, I understand now. Thanks @adriesse I'm sorry too. @thunderfish24 May I suggest we move proposals for customized improvements to `scipy.optimize` subroutines to another issue?

Reply:
I think Bishop method is a sufficient and extremely simple "ground truth" algorithm at the "uncontrolled" terminal voltages that it produces. However, I am sufficiently skeptical that I would want to see a consensus of 2+ algorithms produce the same result (up to some tolerance) at any other voltage values that require an iterative solver, be it `scipy.optimize`'s `fminbnd`, `newton`, or `brentq`, or LambertW with argument overflow protection logic. If this is not established by algorithm consensus, then something like interval analysis on the model residuals would be convincing. I received push back for doing something even less daring with model residual computations in an early commit to #340. At the very least, for the "slow but stable" method, we should make sure all convergence flags are true, and maybe even make assertions on the absolute residuals for the substituted solution (unless the solver has already verified that as some solvers use other stopping criteria).

Reply:
To illustrate brentq() approach more concretely: https://github.com/pvlib/pvlib-python/pull/412